sql: make rewind work correctly with session variables

otan opened the pull request:
Refs: #69391

See individual commits for details.

Release justification: fix to new functionality

otan committed:
sql: remove the use of sessiondata.SessionData as a field

Putting sessiondata.SessionData as a field element can result in obscure
bugs in SET LOCAL, as they may not be using the *top* element of the
stack anymore. Fix this by replacing the fields using
sessiondata.SessionData to use the stack instead.

This is a pre-requisite for the subsequent fix commit.

Release justification: fixes to new functionality.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/69554)
<!-- Reviewable:end -->

rafiss reviewed (changes requested):
some small comments

also, let's get a schema sign-off for the CollectionFactory change

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69554#-:-MiJnavl3LNFvfRuqazJ:b-iekuub)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[-- commits, line 52 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69554#-MiJkfx47WtRCybHOvKP):*
nit: not/now

:P 

though up to you if you wanna skip the release note (but keep it in the commit msg) since the bug never was released
___
*[pkg/sql/catalog/descs/temporary_descriptors.go, line 42 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69554#-MiJkHj5DYwy01jjuPf6:-MiJkHj67bmDJyJufWrw:b-eoysf4) ([raw file](https://github.com/cockroachdb/cockroach/blob/2eca0b20310e27748d9b0a7dbd03e110a1ed5456/pkg/sql/catalog/descs/temporary_descriptors.go#L42)):*
> ```Go
> // sessionData returns the top element of the sessionDataStack used by
> // the current session.
> func (td *temporaryDescriptors) sessionData() *sessiondata.SessionData {
> ```

is this only going to be called from a single goroutine? otherwise we'll end up in a similar race as the default int size one?
___
*[pkg/sql/logictest/testdata/logic_test/txn, line 1099 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69554#-MiJljtiCiU4r98T6fhe:-MiJljtiCiU4r98T6fhf:b-i6eudq) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc9165ff41ff4232383099e0ea8677df433820ec/pkg/sql/logictest/testdata/logic_test/txn#L1099)):*
> ```
> SET SESSION CHARACTERISTICS AS TRANSACTION DEFERRABLE
> 
> # Test retry rewinds correctly.
> ```

cool tests!
___
*[pkg/sql/logictest/testdata/logic_test/txn, line 1109 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69554#-MiJmtwO68JFVxLsMHEN:-MiJmtwO68JFVxLsMHEO:bb0sdxo) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc9165ff41ff4232383099e0ea8677df433820ec/pkg/sql/logictest/testdata/logic_test/txn#L1109)):*
> ```
> BEGIN;
> INSERT INTO rewind_session_test VALUES ('1 day 01:02:03'::interval::string);
> SET intervalstyle = 'iso_8601';
> ```

hmmm this is interesting. it seems like after this fix we're in a middle ground between SET respecting txns sometimes but not all the time. if the txn is retried, the var is reset, but if the txn is just ROLLBACK'd the session var change still goe into effect. i actually don't know what's more "correct"

i feel like it's worth merging this fix

so in that case, maybe we should just go ahead and fix https://github.com/cockroachdb/cockroach/issues/69396 all the way. perhaps we consult product for this?


<!-- Sent from Reviewable.io -->

otan committed:
sql: make rewind work correctly with session variables

Release justification: bug fix for old bug / new features.

Release note (bug fix): Previously, doing SET in a transaction and
having the transaction retry internally could result in the previous
session variable being unused. For example:
```
SET intervalstyle = 'postgres';
BEGIN;
do something with interval -- (1)
SET intervalstyle = 'iso_8601';
do something with interval -- (2)
COMMIT;
```
If the transaction retries at COMMIT, when attempting to re-run (1), we
would have an interval style 'iso_8601' instead of the original
'postgres' value. This has not been resolved.

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69554)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @rafiss)

___
*[pkg/sql/catalog/descs/temporary_descriptors.go, line 42 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69554#-MiJkHj5DYwy01jjuPf6:-MiJqbx63UQOpsqMqVda:bhcs01x) ([raw file](https://github.com/cockroachdb/cockroach/blob/2eca0b20310e27748d9b0a7dbd03e110a1ed5456/pkg/sql/catalog/descs/temporary_descriptors.go#L42)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

is this only going to be called from a single goroutine? otherwise we'll end up in a similar race as the default int size one?
</blockquote></details>

yep single goroutine
___
*[pkg/sql/logictest/testdata/logic_test/txn, line 1109 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69554#-MiJmtwO68JFVxLsMHEN:-MiJqddp1ZfMPWLd0ywY:b-jm1dct) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc9165ff41ff4232383099e0ea8677df433820ec/pkg/sql/logictest/testdata/logic_test/txn#L1109)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

hmmm this is interesting. it seems like after this fix we're in a middle ground between SET respecting txns sometimes but not all the time. if the txn is retried, the var is reset, but if the txn is just ROLLBACK'd the session var change still goe into effect. i actually don't know what's more "correct"

i feel like it's worth merging this fix

so in that case, maybe we should just go ahead and fix https://github.com/cockroachdb/cockroach/issues/69396 all the way. perhaps we consult product for this?


</blockquote></details>

i think this is different -- ROLLBACK implies the session variable is rolled back and is explicitly made by the user, but REWIND is an internal operation which the user might not know happen. it'd be surprising if a user wanted their transaction to only respect the new variable state if we retry a transaction internally.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
lgtm! but let's get schema to be aware of the 2nd commit

ajwerner reviewed (commented):
I'm okay with the 2nd commit but have a suggestion to make it better. 

> Reviewed 2 of 9 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69554#-:-MiKTfFs7NzLraGEOQ0o:b1u9rl4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @otan, and @rafiss)

___
*[pkg/sql/catalog/descs/factory.go, line 46 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/69554#-MiKUNFDBqR3LCf6EZZK:-MiKUNFDBqR3LCf6EZZL:b3c7lx2) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f62602b5a196d6af4e1f6eae7ae84037cf838e/pkg/sql/catalog/descs/factory.go#L46)):*
> ```Go
> 
> // MakeCollection constructs a Collection for the purposes of embedding.
> func (cf *CollectionFactory) MakeCollection(sessionDataStack *sessiondata.Stack) Collection {
> ```

the other option here, rather than exposing the stack directly, is to make an interface that has the parts we need. That's my preference if you're willing. It'll be good overall cleanup. The things we need access to are:

```
SearchPath.GetTemporarySchemaName()
GetTemporarySchemaIDForDb(uint32) uint32
MaybeGetDatabaseForTemporarySchemaID(uint32) (uint32, bool)
```

One way this could go is to have an interface here as the dependency and then implement it in `sql` so we get nice type safety. You'll need to also add the interface to the `EvalContext` in some form so that the distsql code can get its hands on one of these. 

```go
package descs

type TemporarySchemaProvider interface {
    GetTemporarySchemaName() string
    GetTemporarySchemaIDForDB(descpb.ID) descpb.ID
    MaybeGetDatabaseForTemporarySchemaID(descpb.ID) (descpb.ID, bool)
}
```

```go
package sql

type sessionDataTemporarySchemaProvider sessiondata.Stack

var _ descs.TemporarySchemaProvider = (*sessionDataTemporarySchemaProvider)(nil)

// implement the methods based on the top of the stack.
```
___
*[pkg/sql/catalog/descs/temporary_descriptors.go, line 63 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/69554#-MiKTXSf61A9Al9pSkzE:-MiKTXSf61A9Al9pSkzF:b-1l3j1) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f62602b5a196d6af4e1f6eae7ae84037cf838e/pkg/sql/catalog/descs/temporary_descriptors.go#L63)):*
> ```
> 	sd := td.sessionData()
> 	if sd != nil {
> ```
nit: `if sd := td.sessionData(); sd != nil {`


<!-- Sent from Reviewable.io -->

otan committed:
descs: remove the need to import sessiondata.Stack

Replace sessiondata.Stack in the temporaryDescriptors struct with an
interface containing only pertinent information.

Release justification: cleanup change for new feature

Release note: None

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69554)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @otan, and @rafiss)

___
*[pkg/sql/catalog/descs/factory.go, line 46 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/69554#-MiKUNFDBqR3LCf6EZZK:-MiOQAHD9DaI0w36byf8:b63djxd) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f62602b5a196d6af4e1f6eae7ae84037cf838e/pkg/sql/catalog/descs/factory.go#L46)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

the other option here, rather than exposing the stack directly, is to make an interface that has the parts we need. That's my preference if you're willing. It'll be good overall cleanup. The things we need access to are:

```
SearchPath.GetTemporarySchemaName()
GetTemporarySchemaIDForDb(uint32) uint32
MaybeGetDatabaseForTemporarySchemaID(uint32) (uint32, bool)
```

One way this could go is to have an interface here as the dependency and then implement it in `sql` so we get nice type safety. You'll need to also add the interface to the `EvalContext` in some form so that the distsql code can get its hands on one of these. 

```go
package descs

type TemporarySchemaProvider interface {
    GetTemporarySchemaName() string
    GetTemporarySchemaIDForDB(descpb.ID) descpb.ID
    MaybeGetDatabaseForTemporarySchemaID(descpb.ID) (descpb.ID, bool)
}
```

```go
package sql

type sessionDataTemporarySchemaProvider sessiondata.Stack

var _ descs.TemporarySchemaProvider = (*sessionDataTemporarySchemaProvider)(nil)

// implement the methods based on the top of the stack.
```


</blockquote></details>

i gave this a go, couple of things:
* we need to receive a pointer, as the rewind depends on having the stack reference the same place. 
* the impl lives in desc, because it is needed in `sql` and `sql/distsql`. i could move it elsewhere, not sure about tree being the right place.

see the *now* second commit (we moved every commit one down as the first commit got merged to master).

let me know if that's what you had in mind.
___
*[pkg/sql/logictest/testdata/logic_test/txn, line 1099 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69554#-MiJljtiCiU4r98T6fhe:-MiOQjDJEHdN7SVnK1AH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc9165ff41ff4232383099e0ea8677df433820ec/pkg/sql/logictest/testdata/logic_test/txn#L1099)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

cool tests!
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/txn, line 1109 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/69554#-MiJmtwO68JFVxLsMHEN:-MiOQjTYAHRZxYbF_Jxu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc9165ff41ff4232383099e0ea8677df433820ec/pkg/sql/logictest/testdata/logic_test/txn#L1109)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

i think this is different -- ROLLBACK implies the session variable is rolled back and is explicitly made by the user, but REWIND is an internal operation which the user might not know happen. it'd be surprising if a user wanted their transaction to only respect the new variable state if we retry a transaction internally.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ajwerner commented on pkg/sql/catalog/descs/temporary_descriptors.go:
```diff
@@ -24,19 +24,57 @@ import (
+	MaybeGetDatabaseForTemporarySchemaID(descpb.ID) (descpb.ID, bool)
+}
+
+type temporarySchemaProviderImpl struct {
```
In what ways is this better than making the impl a `sessiondata.Stack` for its underlying type?

otan replied:
can't import descpb.ID from there `sessiondata` :D

ajwerner replied:
Like why 

```
type temporarySchemaProviderImpl struct {
   sds *sessiondata.Stack
}
```

and not:

```
type temporarySchemaProviderImpl sessiondata.Stack
```

The latter let's you do everything that the former does and avoids an allocation.

ajwerner replied:
Imagine an sds() method that does the cast.

otan replied:
we need to be a pointer since otherwise the collection in `extraTxnState.collectionFactory` object points to a copy of the stack, which is incorrect.

ajwerner replied:
I'm still not following. Consider:

```go
type temporarySchemaProviderImpl sessiondata.Stack

// NewTemporarySchemaProvider creates a TemporarySchemaProvider.
func NewTemporarySchemaProvider(sds *sessiondata.Stack) TemporarySchemaProvider {
	return (*temporarySchemaProviderImpl)(sds)
}

func (impl *temporarySchemaProviderImpl) sds() *sessiondata.Stack {
    return (*sessionadata.Stack)(impl)
}

// GetTemporarySchemaName implements the TemporarySchemaProvider interface.
func (impl *temporarySchemaProviderImpl) GetTemporarySchemaName() string {
	return impl.sds().Top().TemporarySchemaName
}

// GetTemporarySchemaIDForDB implements the TemporarySchemaProvider interface.
func (impl *temporarySchemaProviderImpl) GetTemporarySchemaIDForDB(id descpb.ID) (descpb.ID, bool) {
	ret, found := impl.sds().Top().GetTemporarySchemaIDForDB(uint32(id))
	return descpb.ID(ret), found
}

// MaybeGetDatabaseForTemporarySchemaID implements the TemporarySchemaProvider interface.
func (impl *temporarySchemaProviderImpl) MaybeGetDatabaseForTemporarySchemaID(
	id descpb.ID,
) (descpb.ID, bool) {
	ret, found := impl.sds().Top().MaybeGetDatabaseForTemporarySchemaID(uint32(id))
	return descpb.ID(ret), found
}
```

otan replied:
ultimately it may make sense to move temp tables out of session data and make it a global state to avoid this stuff (it's not SET LOCAL affected), but that's a bit more than i'm intended to do right now.

otan replied:
ah wait, i failed to understand `type temporarySchemaProviderImpl sessiondata.Stack`

otan replied:
ok done, but had to do this for the casting to work:

```
 // GetTemporarySchemaName implements the TemporarySchemaProvider interface.
 func (impl *temporarySchemaProviderImpl) GetTemporarySchemaName() string {
       return (*sessiondata.Stack)(impl).Top().TemporarySchemaName
 }
```

sorry for the confusion

ajwerner commented on pkg/sql/catalog/descs/temporary_descriptors.go:
```diff
@@ -24,19 +24,57 @@ import (
-	codec       keys.SQLCodec
-	sessionData *sessiondata.SessionData
+	codec                   keys.SQLCodec
+	temporarySchemaProvider TemporarySchemaProvider
```
nit: call call this field something shorter like provider, we're already in the context of temporary stuff.

ajwerner commented on pkg/ccl/changefeedccl/rowfetcher_cache.go:
```diff
@@ -58,7 +58,7 @@ func newRowFetcherCache(
 		codec:      codec,
 		leaseMgr:   leaseMgr,
-		collection: cf.NewCollection(nil /* sessionData */),
+		collection: cf.NewCollection(nil /* sessionDataStack */),
```
nit: TemporarySchemaProvider

ajwerner commented on pkg/sql/catalog/descs/collection_test.go:
```diff
@@ -59,7 +59,7 @@ func TestCollectionWriteDescToBatch(t *testing.T) {
 	db := s0.DB()
 	descriptors := s0.ExecutorConfig().(sql.ExecutorConfig).CollectionFactory.
-		NewCollection(nil /* sessionData */)
+		NewCollection(nil /* sessionDataStack */)
```
nit: TemporarySchemaProvider

ajwerner reviewed (approved):
Collection changes LGTM. Thanks for adding the interface!

otan commented:
bors r=rafiss,ajwerner

otan commented:
bors r=rafiss,ajwerner

otan commented:
bors r=rafiss,ajwerner

merge commit needs release justification, hah
